address.md

```diff
@@ -0,0 +1,303 @@
+
+```go
+type Address interface {
+    New(p byte, payload []byte) Address
+    Encode(network Network, a Adress) string
+    Decode(s string) Address
+    Checksum(a Address) []byte
+    ValidateChecksum(a Address) bool
+}
+```
+##### New()
+
+New returns an Address for the specified protocol encapsulating corresponding payload. New fails for unknown protocol.
+
+```go
+func New(protocol byte, payload []byte) Address {
+	if protocol < SECP256K1 || protocol > BLS {
+		Fatal(ErrUnknownType)
+	}
+    return Address{
+        Protocol: protocol,
+        Payload:  payload,
+    }
+}
+```
+
+##### Encode()
+
+Software encoding a Filecoin address must:
+
+- produce an address encoded to a known network
+- produce an address encoded to a known protocol
+- produce an address with a valid checksum
+
+Encodes an Address as a string, prepending the network prefix, calculating the checksum and encoding the payload and checksum to [base32](https://tools.ietf.org/html/rfc4648).
+
+```go
+func Encode(network string, a Address) string {
+	if network != "f" && network != "t" {
+        Fatal("Invalid Network")
+	}
+	
+    switch a.Protocol {
+    	case SECP256K1:
+    	case Actor:
+    	case BLS:
+    		cksm := Checksum(a)
+    		return network + a.Protocol + base32.Encode(a.Payload + cksm)
+    	case ID:
+        	return network + a.Protocol + base32.Encode(a.Payload)
+        default:
+        	Fatal("invalid address protocol")
+    }
+}
+```
+
+##### Decode()
+
+Software decoding a Filecoin address must:
+* verify the network is a known network.
+* verify the protocol is a number of a known protocol.
+
+* verify the checksum is valid
+
+Decode an Address from a string by, removing the network prefix validating the the address is of a know protocol prefix to theidecoding the payload from base32, and validating the checksum.
+
+```go
+func DecodeString(a string) Address {
+    if len(a) < 4 {
+        Fatal(ErrInvalidLength)
+    }
+    
+    if a[0] != "f" && a[0] != "t" {
+        Fatal(ErrUnknownNetwork)
+    }
+    
+    protocol := a[1]
+    raw := a[2:] 
+    if protocol == ID {
+        return Address{
+            Protocol: protocol,
+            Payload: raw,
+        }
+    }
+    
+    payload = raw[:len(raw)-CksmLen]
+    if protocol == SECP256K1 || protocol == Actor {
+        if len(payload) != 20 {
+            Fatal(ErrInvalidBytes)
+        }
+    }
+    
+    cksm := payload[len(payload)-CksmLen : len(payload)]
+    if !ValidateChecksum(a, cksm) {
+        Fatal(ErrInvalidChecksum)
+    }
+    
+    return Address{
+        Protocol: protocol,
+        Payload: payload,
+    }
+}
+```
+
+##### Checksum()
+
+Checksum produces a byte array of the blake2b-4 checksum of an addresses protocol and payload.
+
+```go
+const CksmLen = 4
+
+func Checksum(a Address) [CksmLen]byte {
+    digest = blake2b(a.Protocol + a.Payload)
+    return digest[:4]
```

Comment:
@frrist this is not blake2b-4, the `4` needs to be passed to the hashing function as an argument, like here: https://github.com/filecoin-project/go-filecoin/blob/6914061e22bb8725182e5496b1286c914b5b9c26/address/address.go#L15

Reply:
Ahh yes, I missed this bit. Seeing as this code isn't expected to compile and is serving here as an example, do we need to add the hash function configuration code? or could we just do something like
`return blake2b4(a.Protocol + a.Payload)`

Reply:
you can do that, and add a comment maybe saying what this is supposed to mean?

Reply:
Sounds good to me, will do that.